src/query/storage/cache/redis_cache.go

```diff
@@ -262,18 +232,155 @@ func WindowGetOrFetch(
+			end := b.End.Unix()
+			prev := idx
+			// Convert millisecond timestamp to second timestamp
+			// Get all samples within the bucket
+			for idx < len(ts.Samples) && ts.Samples[idx].Timestamp/1000 >= start && ts.Samples[idx].Timestamp/1000 < end {
+				idx += 1
+			}
+			// Nothing to add
+			if prev == idx {
+				continue
+			}
+			results[i].PromResult.Timeseries = append(results[i].PromResult.Timeseries,
+				&prompb.TimeSeries{
+					Labels:  ts.Labels,
+					Samples: ts.Samples[prev:idx],
+				},
+			)
+			// No need to continue
+			if idx == len(ts.Samples) {
+				break
+			}
+		}
+	}
+	cache.Set(buckets, results, BucketKeyPrefix)
+}
+
+// Function that checks whether the query data is in Redis using a sliding window
+// If it is not, it gets the result from M3DB
+//
+// Currently, this is done by splitting data into buckets
+// and checking if we have enough buckets that we can retrieve from Redis to recreate the result
+func BucketWindowGetOrFetch(
+	ctx context.Context,
+	st storage.Storage,
+	fetchOptions *storage.FetchOptions,
+	q *storage.FetchQuery,
+	cache *RedisCache,
+) (storage.PromResult, error) {
+	if cache == nil || !EnableCache {
+		return st.FetchProm(ctx, q, fetchOptions)
+	}
+	queryRange := int64(q.End.Sub(q.Start).Seconds())
+	// If size is <= bucket size, use simple caching
+	if queryRange <= int64(BucketSize.Seconds()) {
+		return WindowGetOrFetch(ctx, st, fetchOptions, q, cache)
+	}
+
+	last_start := q.End.Truncate(BucketSize)
+	// If we have more than one bucket
+	if q.Start.Before(last_start) {
+		// Split into not-last and last
+		exclude_last_query := &storage.FetchQuery{
+			TagMatchers: q.TagMatchers,
+			Start:       q.Start,
+			End:         last_start,
+			Interval:    q.Interval,
+		}
+		last_query := &storage.FetchQuery{
+			TagMatchers: q.TagMatchers,
+			Start:       last_start,
+			End:         q.End,
+			Interval:    q.Interval,
+		}
+		buckets := splitQueryToBuckets(exclude_last_query, BucketSize)
+		// Check if we have to make a request to M3DB, in which case we just ask for all
+		// This is so we don't have to convert if we need to ask M3DB for all of it anyways
+		count := cache.Check(buckets, BucketKeyPrefix)
+		cache.logger.Info("num_hit", zap.Int("check", count), zap.Int("length", len(buckets)))
+		if count < len(buckets)-1 || count == 0 {
+			cache.logger.Info("cache miss", zap.String("key", KeyEncode(q, BucketKeyPrefix)))
+			res, err := st.FetchProm(ctx, q, fetchOptions)
+			cache.cacheMetrics.CacheMetricsMiss(res)
+			if err == nil {
+				cache.SetAsBuckets(&res, buckets)
+			}
+			return res, err
+		}
+
+		results := cache.Get(buckets, BucketKeyPrefix)
+		cnt := 0
+		// Check again after getting just in case some of the keys disappeared so we don't make many M3DB requests
+		for _, r := range results {
+			if r == nil {
+				cnt++
+				if cnt > 1 {
+					cache.logger.Info("cache miss", zap.String("key", KeyEncode(q, BucketKeyPrefix)))
+					res, err := st.FetchProm(ctx, q, fetchOptions)
```

Comment:
If it is tricky we can keep it simple, two questions:
1. how frequently if middle of bucket cache miss?
2. how hard if this cache miss and we only query this bucket instead of full range?

Reply:
Middle of the bucket doesn't miss very often but some of buckets at the beginning or end can miss depending on how frequently the rule is evaluated by different time windows. Right now if it's just one bucket, we do only query this bucket instead of the full range